Mathlib/GroupTheory/GroupAction/Primitive.lean

```diff
@@ -0,0 +1,365 @@
+/-
+Copyright (c) 2022 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+import Mathlib.GroupTheory.MaximalSubgroups
+import Mathlib.GroupTheory.GroupAction.Blocks
+import Mathlib.GroupTheory.GroupAction.Transitive
+import Mathlib.Data.Setoid.Partition
+
+/-!
+# Primitive actions
+
+## Definitions
+
+- `IsPreprimitive G X`
+a structure that says that the action of a type `G`
+on a type `X` (defined by an instance `SMul G X`) is *preprimitive*,
+namely, it is pretransitive and the only blocks are ⊤ and subsingletons.
+(The pretransitivity assumption is essentially trivial,
+because orbits are blocks, unless the action itself is trivial.)
+
+The notion which is introduced in classical books on group theory
+is restricted to `mul_action` of groups.
```

Comment:
```suggestion
is restricted to `MulAction` of groups.
```

Reply:
It would be nice to have some examples that satisfy each of the weaker typeclasses while not satisfying the stronger ones, as well as some explanation why these notions are interesting.

Reply:
I understand. I can easily give examples in the docstring. It might be more difficult to formalize them right now (for example, can we state the action of SO(3,R) on the sphere in mathlib? It is transitive, primitive, but not 2-transitive. It is in my plans to formalize those actions — for any quadratic form — and to prove the related results. Anisotropic forms are unfortunately more difficult.)